1-Draft/RFCnnnn-ForEach-Parallel-Cmdlet.md

```diff
@@ -0,0 +1,180 @@
+---
+RFC: RFCnnnn
+Author: Paul Higinbotham
+Status: Draft
+SupercededBy: N/A
+Version: 1.0
+Area: Engine
+Comments Due: July 18, 2019
+Plan to implement: Yes
+---
+
+# PowerShell ForEach-Object -Parallel Cmdlet
+
+This RFC proposes a new parameter set for the existing ForEach-Object cmdlet to parallelize script block executions, instead of running them sequentially as it does now.  
+
+## Motivation
+
+    As a PowerShell User,
+    I can execute foreach-object piped input in script blocks running in parallel threads, either synchronously or asynchronously, while limiting the number of threads running at a given time.
+
+## Specification
+
+A new `-Parallel` parameter set will be added to the existing ForEach-Object cmdlet that supports running piped input concurrently in a provided script block.  
+
+- `-Parallel` parameter takes a script block that is executed in parallel for each piped input variable
+
+- `-ThrottleLimit` parameter takes an integer value that determines the maximum number of script blocks running at the same time
+
+- `-TimeoutSeconds` parameter takes an integer that specifies the maximum time to wait for completion before the command is aborted
+
+- `-AsJob` parameter switch indicates that a job is returned, which represents the command running asynchronously
+
+The 'ForEach-Object -Parallel' command will return only after all piped input have been processed.
+Unless the '-AsJob' switch is used, in which case a job object is returned immediately that monitors the ongoing execution state and collects generated data.

```

Comment:
33 and 34 need to be clarified. I've run the following code on a nightly:

```powershell
function foo {foreach ($i in 1..5) {get-random -min 1 -max 10; sleep (get-random -max 5 -min 1)}}

# This appears to stream as each element of the pipeline is processed
foo | foreach-object -parallel {$_}

# This blocks return of a job object until all elements of the pipeline have been processed
foo | foreach-object -parallel {$_} -asjob
```

Would you agree with something like this? 
"Like all cmdlets, `ForEach-Object -Parallel -AsJob` must receive all input (including elements of a pipeline) before it can return a job object.
If you don't use `-AsJob`, then input is streamed to the console."